staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/client.go

```diff
@@ -0,0 +1,212 @@
+		return nil, err
+	}
+
+	return api.NewClient(vaultConfig)
+}
+
+// Get token by login and set the value to api.Client.
+func (c *clientWrapper) refreshToken(config *EnvelopeConfig) error {
+
+	switch {
+	case config.ClientCert != "" && config.ClientKey != "":
+		token, err := c.tlsToken(config)
+		if err != nil {
+			return err
+		}
+		c.client.SetToken(token)
+	case config.RoleID != "":
+		token, err := c.appRoleToken(config)
+		if err != nil {
+			return err
+		}
+		c.client.SetToken(token)
+	default:
+		// configuration has already been validated, flow should not reach here
+		err := fmt.Errorf("The Vault authentication configuration is invalid")
+		return err
+	}
+
+	return nil
+}
+
+func (c *clientWrapper) tlsToken(config *EnvelopeConfig) (string, error) {
+	resp, err := c.client.Logical().Write(c.authPath+"cert/login", nil)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) appRoleToken(config *EnvelopeConfig) (string, error) {
+	data := map[string]interface{}{
+		"role_id":   config.RoleID,
+		"secret_id": config.SecretID,
+	}
+	resp, err := c.client.Logical().Write(c.authPath+"approle/login", data)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Auth.ClientToken, nil
+}
+
+func (c *clientWrapper) decrypt(keyName string, cipher string) (string, error) {
+	var result string
+
+	data := map[string]string{"ciphertext": cipher}
+	resp, err := c.request(c.decryptPath+keyName, data)
+	if err != nil {
+		return result, err
+	}
+
+	result, ok := resp.Data["plaintext"].(string)
+	if !ok {
+		return result, fmt.Errorf("failed type assertion of vault decrypt response type: %v to string", reflect.TypeOf(resp.Data["plaintext"]))
+	}
+
+	return result, nil
+}
+
+func (c *clientWrapper) encrypt(keyName string, plain string) (string, error) {
+	var result string
+
+	data := map[string]string{"plaintext": plain}
+	resp, err := c.request(c.encryptPath+keyName, data)
+	if err != nil {
+		return result, err
+	}
+
+	result, ok := resp.Data["ciphertext"].(string)
+	if !ok {
+		return result, fmt.Errorf("failed type assertion of vault encrypt response type: %v to string", reflect.TypeOf(resp.Data["ciphertext"]))
+	}
+
+	return result, nil
+}
+
+// This request check the response status code. If get code 403, it sets forbidden true.
+func (c *clientWrapper) request(path string, data interface{}) (*api.Secret, error) {
+	req := c.client.NewRequest("POST", path)
+	if err := req.SetJSONBody(data); err != nil {
+		return nil, err
+	}
+
+	resp, err := c.client.RawRequest(req)
+
+	if resp != nil {
+		defer resp.Body.Close()
+		if resp.StatusCode == 403 {
+			return nil, &forbiddenError{err: err}
+		}
+
+		if resp.StatusCode == 200 {
+			secret, err := api.ParseSecret(resp.Body)
+			if err != nil {
+				return nil, err
+			}
+			return secret, nil
+		}
+		return nil, fmt.Errorf("Unexpected response code: %v received for POST request to %v", resp.StatusCode, path)
```

Comment:
errors should start with a lowercase (this goes for other errors like below): https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
fixed